### PR TITLE
Guard MacroConfig reads and writes

### PR DESCRIPTION
**Pull request by stefanpenner**
Add `finalization` concept to MacroConfigs to guard against reading
before writes are done or writing while reads are possible.

The timing ends up being:

1. MacroConfig is writable (not readable)
2. MacroConfig is finalized, and becomes readable but no longer writable
  In general the goal here was allow config writing until just before `treeFor*` hooks are invoked.

  * For Classic builds: once `app.toTree` is invoked, the MacosConfig is finalized, becoming frozen but readonable
  * For Embroider builds: about the same time but via the `default-pipeline` the MacrosConfig  is finalized, becoming frozen but readonable

For the classic app builds the `toTree` usage here is a POC, if we continue down this path I don't mind adding the corresponding public API to ember-cli.

--- 

paired with @thoov

### CODE REVIEW

File: packages/core/src/app.ts [hunk 1]
```diff
@@ -259,9 +264,7 @@ export class AppBuilder<TreeNames> {
     private app: Package,
     private adapter: AppAdapter<TreeNames>,
     private options: Required<Options>
-  ) {
-    MacrosConfig.shared().setOwnConfig(__filename, { active: true });
-  }
+  ) {}
```
Comment (thoov) [inline packages/core/src/app.ts]: Can we remove the constructor completely?

File: packages/macros/src/ember-addon-main.ts [hunk 2]
```diff
@@ -43,6 +56,7 @@ export = {
           },
         });
       });
+
```
Comment (thoov) [inline packages/macros/src/ember-addon-main.ts]: Remove newline

File: packages/macros/src/macros-config.ts [hunk 3]
```diff
@@ -59,9 +65,10 @@ export default class MacrosConfig {
   }
 
   private internalSetConfig(fromPath: string, packageName: string | undefined, config: unknown) {
-    if (this.cachedUserConfigs) {
-      throw new Error(`attempted to set config after we have already emitted our config`);
+    if (this._configWritable === false) {
```
Comment (thoov) [inline packages/macros/src/macros-config.ts]:
```js
if (!this._configWritable) {
```

File: packages/macros/src/macros-config.ts [hunk 4]
```diff
@@ -161,9 +174,10 @@ export default class MacrosConfig {
   // this exists because @embroider/compat rewrites and moves v1 addons, and
   // their macro configs need to follow them to their new homes.
   packageMoved(oldPath: string, newPath: string) {
-    if (this.cachedUserConfigs) {
-      throw new Error(`attempted to call packageMoved after we have already emitted our config`);
+    if (this._configWritable === false) {
```
Comment (thoov) [inline packages/macros/src/macros-config.ts]:
```js
if (this._configWritable === false) {
```

File: packages/macros/tests/babel/helpers.ts [hunk 5]
```diff
@@ -4,14 +4,17 @@ import { allBabelVersions as allBabel, runDefault } from '@embroider/test-suppor
...
 export { runDefault };
-
 export function allBabelVersions(createTests: (transform: (code: string) => string, config: MacrosConfig) => void) {
   let config: MacrosConfig;
 
   describe('without presets', function() {
     allBabel({
       babelConfig() {
         config = new MacrosConfig();
+        if (createTests.length < 2) {
```
Comment (thoov) [inline packages/macros/tests/babel/helpers.ts]: We should probably add a comment about this

File: packages/macros/src/ember-addon-main.ts [hunk 6]
```diff
@@ -25,8 +25,21 @@ export = {
     let babelOptions = (parentOptions.babel = parentOptions.babel || {});
     let babelPlugins = (babelOptions.plugins = babelOptions.plugins || []);
     babelPlugins.unshift(MacrosConfig.shared().babelPluginConfig(source));
+
+    // TODO: LOL MEGA HACK
```
Comment (thoov) [inline packages/macros/src/ember-addon-main.ts]: We should discuss this

**Comment by thoov** [comment]
Do we want to add the ability to turn this into warnings?

File: packages/core/src/app.ts (hunk 1 above)
Comment (stefanpenner) [reply packages/core/src/app.ts]: ya

File: packages/macros/src/ember-addon-main.ts (hunk 6 above)
Comment (stefanpenner) [reply packages/macros/src/ember-addon-main.ts]: yes we must

File: packages/macros/src/ember-addon-main.ts (hunk 2 above)
Comment (thoov) [reply packages/macros/src/ember-addon-main.ts]: Fixed

File: packages/macros/src/macros-config.ts (hunk 3 above)
Comment (thoov) [reply packages/macros/src/macros-config.ts]: Fixed

File: packages/macros/src/macros-config.ts (hunk 4 above)
Comment (thoov) [reply packages/macros/src/macros-config.ts]: Fixed

File: packages/macros/src/ember-addon-main.ts (hunk 6 above)
Comment (stefanpenner) [reply packages/macros/src/ember-addon-main.ts]: If we decide this is the correct timing, I don't mind adding the appropriate hook to ember-cli proper, and using this as a temp polyfil.

cc @ef4

**Comment by stefanpenner** [comment]
> Do we want to add the ability to turn this into warnings?

Let's wait, and let a use-case drive that. I would prefer to keep it strict, as other "downstream" issues may be pretty gnarly to debug.

File: packages/macros/tests/babel/helpers.ts (hunk 5 above)
Comment (stefanpenner) [reply packages/macros/tests/babel/helpers.ts]: ya

Comment (stefanpenner) [reply packages/macros/tests/babel/helpers.ts]: doing that now.

File: packages/macros/src/ember-addon-main.ts [hunk 7]
```diff
@@ -3,28 +3,46 @@ import { MacrosConfig } from '.';
...
+    //
+    // In general, we do not condone this type of monkey patching and our plan
+    // is if this indeed remains needed, to add the appropriate public API in
+    // ember-cli. One interesting tid-bit, once users use embroider primarily
+    // this hack will have no impact, and then adding the API to ember-cli
+    // may be wasted. We will see how this plays out, and do the appropriate
+    // thing.
+    const originalToTree = parent.toTree;
+    parent.toTree = function() {
+      debugger;
```
Comment (kratiahuja) [inline packages/macros/src/ember-addon-main.ts]: We should remove this debugger

File: packages/macros/src/macros-config.ts [hunk 8]
```diff
@@ -9,38 +9,70 @@ export type Merger = (configs: unknown[]) => unknown;
...
+  // }
+
+  static for(key: any): MacrosConfig {
+    let config = CONFIGS.get(key);
+    if (config) {
+      return config;
     }
-    return localSharedState;
+    COUNTER;
+    debugger;
```
Comment (kratiahuja) [inline packages/macros/src/macros-config.ts]: Ditto.

**Pull request merged**